presto-main/src/main/java/com/facebook/presto/operator/ParametricImplementations.java

```diff
@@ -57,4 +67,72 @@ public ParametricImplementations(Map<Signature, T> exactImplementations, List<T>
     {
         return specializedImplementations;
     }
+
+    public Signature getSignature()
+    {
+        return signature;
+    }
+
+    public static <T extends ParametricImplementation> Builder<T> builder()
+    {
+        return new Builder<>();
+    }
+
+    public static final class Builder<T extends ParametricImplementation>
+    {
+        private final ImmutableMap.Builder<Signature, T> exactImplementations = ImmutableMap.builder();
+        private final ImmutableList.Builder<T> specializedImplementations = ImmutableList.builder();
+        private final ImmutableList.Builder<T> genericImplementations = ImmutableList.builder();
+
+        private Builder() {}
+
+        public ParametricImplementations<T> build()
+        {
+            Map<Signature, T> exactImplementations = this.exactImplementations.build();
+            List<T> specializedImplementations = this.specializedImplementations.build();
+            List<T> genericImplementations = this.genericImplementations.build();
+            return new ParametricImplementations<T>(exactImplementations,
+                    specializedImplementations,
+                    genericImplementations,
+                    determineGenericSignature(exactImplementations, specializedImplementations, genericImplementations));
+        }
+
+        public void addImplementation(T implementation)
+        {
+            if (implementation.getSignature().getTypeVariableConstraints().isEmpty()
+                    && implementation.getSignature().getArgumentTypes().stream().noneMatch(TypeSignature::isCalculated)
+                    && !implementation.getSignature().getReturnType().isCalculated()) {
+                exactImplementations.put(implementation.getSignature(), implementation);
+            }
+            else if (implementation.hasSpecializedTypeParameters()) {
+                specializedImplementations.add(implementation);
+            }
+            else {
+                genericImplementations.add(implementation);
+            }
+        }
+
+        private static <T extends ParametricImplementation> Signature determineGenericSignature(
+                Map<Signature, T> exactImplementations,
+                List<T> specializedImplementations,
+                List<T> genericImplementations)
+        {
+            if ((specializedImplementations.size() + genericImplementations.size()) == 0) {
```

Comment:
`specializedImplementations.size() + genericImplementations.size() == 0`

That pair of parenthesis is excessive